app/get_proc_arg.go

```diff
@@ -0,0 +1,225 @@
+package app
+
+import (
+	"fmt"
+	"github.com/pkg/errors"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+
+	"github.com/rancher/longhorn-manager/types"
+)
+
+const (
+	DetectPodMaxPolls = 120
+
+	ArgFlexvolumePluginDir = "volume-plugin-dir"
+	ArgKubeletRootDir      = "root-dir"
+
+	DefaultFlexvolumeDir  = "/usr/libexec/kubernetes/kubelet-plugins/volume/exec/"
+	DefaultKubeletRootDir = "/var/lib/kubelet"
+
+	DetectFlexvolumeName     = "discover-flexvolume-dir"
+	DetectKubeletRootDirName = "discover-kubelet-root-dir"
+
+	DetectKubeletArgScript = `
+    find_kubelet_proc() {
+      for proc in $(find /proc -maxdepth 1 -type d 2>/dev/null); do
+        if [ ! -f $proc/cmdline ]; then
+          continue
+        fi
+        if [[ "$(cat $proc/cmdline | tr '\000' '\n' | head -n1 | tr '/' '\n' | tail -n1)" == "kubelet" ]]; then
+          echo $proc
+        fi
+      done
+    }
+    get_kubelet_arg_dir() {
+      proc=$(find_kubelet_proc)
+      if [ "$proc" != "" ]; then
+        path=$(cat $proc/cmdline | tr '\000' '\n' | grep %s | tr '=' '\n' | tail -n1)
+        if [ "$path" == "" ]; then
+          echo "No arg root-dir in proc kubelet. Will use default root-dir."
+          echo %s
+        else
+          echo $path
+        fi
+      fi
+    }
+	get_kubelet_arg_dir
+	`
+	DetectK3SDataDirScript = `
+    find_k3s_proc() {
+      for proc in $(find /proc -maxdepth 1 -type d 2>/dev/null); do
+        if [ ! -f $proc/cmdline ]; then
+          continue
+        fi
+        if [[ $(cat $proc/cmdline | tr '\000' '\n' | sed -n '1p') == "k3s" ]]; then
+          proc_name=$(cat $proc/cmdline | tr '\000' '\n' | sed -n '2p')
+          if [[ "$proc_name"  == "server" || "$proc_name" == "agent" ]]; then
+            echo $proc
+          fi
+        fi
+      done
+    }
+    get_k3s_data_dir() {
+      proc=$(find_k3s_proc)
+      if [ "$proc" != "" ]; then
+        path=$(cat $proc//cmdline | tr '\000' '\n' |  grep 'data-dir' -A 1 | tail -n1)
+        if [ "$path" == "" ]; then
+          path=$(cat $proc//cmdline | tr '\000' '\n' |  grep '\^-d$' -A 1 | tail -n1)
+        fi
+        if [ "$path" == "" ]; then
+          echo "No arg data-dir in proc k3s. Will use default data-dir."
```

Comment:
Is this echo'ing two lines when falling back to default, with this comment then getting parsed as the directory?

Reply:
This echoing will work only when Longhorn has found proc `k3s` but doesn't detect the argument `data-dir` set in its cmdline. If Longhorn cannot find proc `k3s` or `kubelet`, it will directly error out.

Reply:
Yep, so this line is being echo'd along with the default k3s path on the next line. However, you're expecting the output from this pod to just be the path in `discoverProcArg`, so this comment is getting parsed as the path and failing.

Reply:
No. 

The echo'd `"No arg data-dir in proc k3s. Will use default data-dir."` will be processed here https://github.com/rancher/longhorn-manager/blob/de3962d5a0b99b70f440dc4cef2c43773097026f/app/get_proc_arg.go#L220

And this issue is caused by https://github.com/rancher/longhorn-manager/blob/de3962d5a0b99b70f440dc4cef2c43773097026f/app/get_proc_arg.go#L46

The script is expecting `k3s server` without prefix `/usr/local/bin/`, Hence it cannot find proc `k3s` here and these echo'es won't get triggered . I am fixing and testing it right now.

Thank you!